metadata-ingestion/src/datahub/ingestion/source/redash.py

```diff
@@ -0,0 +1,466 @@
+        description = self._get_dashboard_description_from_widgets(widgets)
+        chart_urns = self._get_dashboard_chart_urns_from_widgets(widgets)
+
+        dashboard_info = DashboardInfoClass(
+            description=description,
+            title=title,
+            charts=chart_urns,
+            lastModified=last_modified,
+            dashboardUrl=dashboard_url,
+            customProperties={},
+        )
+        dashboard_snapshot.aspects.append(dashboard_info)
+
+        return dashboard_snapshot
+
+    def _emit_dashboard_mces(self) -> Iterable[MetadataWorkUnit]:
+        current_dashboards_page = 0
+        skip_draft = self.config.skip_draft
+
+        # we will set total dashboards to the actual number after we get the response
+        total_dashboards = PAGE_SIZE
+
+        while (
+            current_dashboards_page * PAGE_SIZE <= total_dashboards
+            and current_dashboards_page < self.api_page_limit
+        ):
+            dashboards_response = self.client.dashboards(
+                page=current_dashboards_page + 1, page_size=PAGE_SIZE
+            )
+            total_dashboards = dashboards_response.get("count") or 0
+            current_dashboards_page += 1
+
+            logger.info(f"/api/dashboards on page {current_dashboards_page}")
+
+            for dashboard_response in dashboards_response["results"]:
+
+                dashboard_name = dashboard_response["name"]
+
+                if (not self.config.dashboard_patterns.allowed(dashboard_name)) or (
+                    skip_draft and dashboard_response["is_draft"]
+                ):
+                    self.report.report_dropped(dashboard_name)
+                    continue
+
+                # Continue producing MCE
+                dashboard_slug = dashboard_response["slug"]
+                dashboard_data = self.client.dashboard(dashboard_slug)
+                dashboard_snapshot = self._get_dashboard_snapshot(dashboard_data)
+                mce = MetadataChangeEvent(proposedSnapshot=dashboard_snapshot)
+                wu = MetadataWorkUnit(id=dashboard_snapshot.urn, mce=mce)
+                self.report.report_workunit(wu)
+
+                yield wu
+
+    def _get_chart_type_from_viz_data(self, viz_data):
+        """
+        https://redash.io/help/user-guide/visualizations/visualization-types
+        Redash has multiple visualization types. Chart type is actually Plotly.
+        So we need to check options returned by API, which series type is being used.
+        """
+        viz_type = viz_data.get("type", "")
+        viz_options = viz_data.get("options", {})
+        globalSeriesType = viz_options.get("globalSeriesType", "")
+        report_key = f"redash-chart-{viz_data['id']}"
+
+        # handle Plotly chart types
+        if viz_type == "CHART":
+            chart_type = PLOTLY_CHART_MAP.get(
+                globalSeriesType, DEFAULT_VISUALIZATION_TYPE
+            )
+            if not chart_type:
```

Comment:
If the default is `DEFAULT_VISUALIZATION_TYPE`, then is it really possible for `chart_type` to be falsy?

Reply:
Yes, it possible. That's why we need to add more classes in `ChartTypeClass`. However, I don't know how to add this, since as far as I can understand `schema_classes.py` is generated from JAVA.

Reply:
Improve the logic of chart type detection and report warning for missing chart types

https://github.com/linkedin/datahub/blob/13210f991b7a9a5abd2fb808ca6c1c3846cf361d/metadata-ingestion/src/datahub/ingestion/source/redash.py#L341-L347

Tested example:
```bash
 'warnings': {'redash-chart-11542': ['ChartTypeClass for Redash Visualization Type=COHORT is missing. Setting to TABLE'],
              'redash-chart-11544': ['ChartTypeClass for Redash Visualization Type=COHORT is missing. Setting to TABLE'],
              'redash-chart-11712': ['ChartTypeClass for Redash Visualization Type=COHORT is missing. Setting to TABLE'],
              'redash-chart-11870': ['ChartTypeClass for Redash Visualization Type=COHORT is missing. Setting to TABLE'],
              'redash-chart-12107': ['ChartTypeClass for Redash Visualization Type=COHORT is missing. Setting to TABLE'],
              'redash-chart-14050': ['ChartTypeClass for Redash Visualization Type=COHORT is missing. Setting to TABLE'],
```